src/wallet/scriptpubkeyman.cpp

```diff
@@ -1717,42 +1717,190 @@ std::unordered_set<CScript, SaltedSipHasher> LegacyDataSPKM::GetScriptPubKeys()
+        case TxoutType::WITNESS_V1_TAPROOT:
+        // These are only spendable if the witness script is also spendable as a scriptPubKey
+        // We will check these later after "spks" has been updated with the computed output scripts from mapScripts.
+        case TxoutType::WITNESS_V0_SCRIPTHASH:
+        // For P2SH to be spendable, we need to have the redeemScript, and if we have it, it will be handled when
+        // this loop gets to it. A P2SH script itself cannot be nested in anything, so we can skip them.
+        case TxoutType::SCRIPTHASH:
+            break;
+        // Any P2PK or P2PKH scripts found in mapScripts can be spent as P2SH-P2PK or P2SH-P2PKH respectively,
+        // if we have the private key.
+        // Since all private keys were iterated earlier and their corresponding P2PK and P2PKH scripts inserted
+        // to "spks", we can simply check whether this P2PK or P2PKH script is in "spks" to determine spendability.
+        case TxoutType::PUBKEY:
+        case TxoutType::PUBKEYHASH:
+            if (spks.count(script) > 0) {
                 spks.insert(GetScriptForDestination(ScriptHash(script)));
             }
-            // For segwit scripts, we only consider them spendable if we have the segwit spk
-            int wit_ver = -1;
-            std::vector<unsigned char> witprog;
-            if (script.IsWitnessProgram(wit_ver, witprog) && wit_ver == 0) {
+            break;
+        // P2WPKH scripts are only spendable if we have the private key.
+        case TxoutType::WITNESS_V0_KEYHASH:
+        {
+            CKeyID key_id{uint160(solutions[0])};
+            CPubKey pubkey;
+            if (GetPubKey(key_id, pubkey) && pubkey.IsCompressed() && HaveKey(key_id)) {
                 spks.insert(script);
+                // Also insert P2SH-P2WPKH output script
+                spks.insert(GetScriptForDestination(ScriptHash(script)));
             }
-        } else {
-            // Multisigs are special. They don't show up as ISMINE_SPENDABLE unless they are in a P2SH
-            // So check the P2SH of a multisig to see if we should insert it
-            std::vector<std::vector<unsigned char>> sols;
-            TxoutType type = Solver(script, sols);
-            if (type == TxoutType::MULTISIG) {
-                CScript ms_spk = GetScriptForDestination(ScriptHash(script));
-                if (IsMine(ms_spk) != ISMINE_NO) {
-                    spks.insert(ms_spk);
-                }
+            break;
+        }
+        // Multisig scripts are spendable if they are inside of a P2SH or P2WSH, and we have all of the private keys.
+        // Bare multisigs are never considered spendable
+        case TxoutType::MULTISIG:
+        {
+            Assert(solutions.size() >= 2);
+            std::vector<std::vector<unsigned char>> keys(solutions.begin() + 1, solutions.begin() + solutions.size() - 1);
+            if (!HaveKeys(keys, *this)) {
+                break;
             }
```

Comment:
At this point, the witness program, witness script, or redeemscript is in our mapScript, and we might have enough keys for the quorum but just not all of the keys.

Why does this require all the keys to be present in the wallet rather than just the threshold?

Reply:
To maintain strict compatibility with `LegacyScriptPubKeyMan::IsMine()`. I don't think we should change this behavior in this PR.